lib/util/http2_rpc.c

```diff
@@ -44,6 +45,89 @@ typedef struct http2_rpc_data {
 	size_t data_length;
 } http2_rpc_data;
 
+static int decompress_grpc_response(http2_rpc_handle *hd)
+{
+	if (hd->bytes_written < 7) {
+		return 1;
+	}
+
+	// Check if gRPC response starts with compression flag and has gzip magic bytes
+	if ((unsigned char)hd->output_buffer[0] != 0x01) {
+		// No compression flag found, data appears uncompressed
+		return 1;
+	}
+
+	// Check for gzip magic bytes
+	if ((unsigned char)hd->output_buffer[5] != 0x1f || (unsigned char)hd->output_buffer[6] != 0x8b) {
+		// No gzip magic bytes found at offset
+		return -1;
+	}
+
+	z_stream stream;
+	(void)memset(&stream, 0, sizeof stream);
+
+	int ret = inflateInit2(&stream, 15 + 32); // Auto-detect gzip/deflate
+	if (ret != Z_OK) {
+		return -1;
+	}
+
+	stream.avail_in = (uInt)(hd->bytes_written - 5);
+	stream.next_in = (Bytef *)(hd->output_buffer + 5);
+
+	size_t decompressed_len = 0;
+	size_t decompressed_cap = 4096;
+	char *decompressed_data = malloc(decompressed_cap);
+	if (!decompressed_data) {
+		inflateEnd(&stream);
+		return -1;
+	}
+
+	do {
+		const size_t min_avail_out = 1024;
+		if (decompressed_cap - decompressed_len < min_avail_out) {
+			decompressed_cap += min_avail_out;
+			void *new_buffer = realloc(decompressed_data, decompressed_cap);
+			if (!new_buffer) {
+				free(decompressed_data);
+				inflateEnd(&stream);
+				return -1;
+			}
+			decompressed_data = new_buffer;
+		}
+
+		size_t avail_out = decompressed_cap - decompressed_len;
+		stream.avail_out = (uInt)avail_out;
+		stream.next_out = (Bytef *)(decompressed_data + decompressed_len);
+
+		ret = inflate(&stream, Z_NO_FLUSH);
+		if (ret == Z_NEED_DICT) {
+			ret = Z_DATA_ERROR;
+		}
+		if (ret == Z_STREAM_ERROR || ret == Z_DATA_ERROR || ret == Z_MEM_ERROR) {
+			free(decompressed_data);
+			(void)inflateEnd(&stream);
+			return -1;
+		}
+
+		decompressed_len += avail_out - stream.avail_out;
+	} while (stream.avail_out == 0);
+
+	if (ret != Z_STREAM_END) {
+		free(decompressed_data);
+		(void)inflateEnd(&stream);
+		return -1;
+	}
+
+	(void)inflateEnd(&stream);
```

Comment:
For easier cleanup would suggest to add the following cleanup labels:
```c
cleanup_decompressed_data:
  free(decompressed_data);

cleanup_stream:
  (void)inflateEnd(&stream);

if (ret != 0) return ret;
```

Then you can jump into these labels when an error occurs.

Reply:
Opted against it in this case. `decompressed_data` doesn't need to be free'ed in all cases and the cleanup logic got therefore complicated too. With this insight, I think the current control flow is easy enough to follow as is. What's your opinion?